undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPIAuthenticationMechanism.java

```diff
@@ -135,7 +135,7 @@ public AuthenticationMechanismOutcome authenticate(final HttpServerExchange exch
                 cache = Boolean.valueOf((String)registerObj);
             }
             sc.authenticationComplete(authenticatedAccount, authType, cache);
-        } else if (isValid && authenticatedAccount == null && !isMandatory(requestContext)) {
+        } else if (!isMandatory(requestContext)) {
             outcome = AuthenticationMechanismOutcome.NOT_ATTEMPTED;
```

Comment:
The problem with the spec is that it transfers too much responsibility to the external AuthModule implementions. In some cases it is the module's responsibility to establish the container representation of the unauthenticated identity. Here's the full paragraph that describes the non mandatory auth access:

"When validateRequest is called before the service invocation on a module that was initialized with an optional requestPolicy (that is, requestPolicy.isMandatory() returns false), the module should attempt to satisfy the request authentication policy, but it must do so without initiating additional message exchanges or interactions involving the client. Independent of whether the authentication policy is satisfied, the module may return AuthStatus.SUCCESS. If the module returns AuthStatus.SUCCESS (and the authentication policy was satisfied), the module (or its context) must employ a CallerPrincipalCallback as described above. If the authentication policy was not satisfied, and yet the module chooses to return AuthStatus.SUCCESS, the module (or its context) must use a CallerPrincipalCallback to establish the container’s representation of the unauthenticated caller within the clientSubject. If the module determines that an invalid or incomplete security context was used to secure the request, then the module may return AuthStatus.SEND_FAILURE, AuthStatus.SEND_CONTINUE, or throw an AuthException. If the module throws an AuthException, or returns any value other that AuthStatus.SUCCESS, the runtime must NOT proceed to the service invocation. The runtime must process an AuthException as described above for a request with a mandatory requestPolicy. The runtime must process any return value other than AuthStatus.SUCCESS as it would be processed if it were returned for a request with a mandatory requestPolicy."

I assume that when it talks about the module being able to satisfy the auth policy it means that you have the proactive auth scenario (a credential is available to authenticate so you do it as you described to have a user available in logs, etc). If the module can't satisfy the policy it can chose between just using an unauthenticated identity and reporting an error and if it does the latter then the request should never reach the resource.